Support for log level understanding and control

ktf opened the pull request:
This allows controlling the log level from the GUI and
colorize things differently depending on it. It also extends
the GUI to allow changing the log level dynamically, on a
per Device level.

Moreover, in case an error is encountered in any of the Data Processors, the application
exits with 1 rather than 0. This should allow us to spot
test that while not crashing, do not behave correctly.

alibuild commented:
Error while checking build/o2checkcode/o2 for 0e2adac8e237c59b58c9f59802920d1c8ca1eae0:
```
sw/BUILD/O2-latest/log
[ 37%] Building CXX object Utilities/QC/QCMerger/CMakeFiles/runQCMergerDevice.dir/src/runMerger.cxx.o
[ 37%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/DeviceSpec.cxx.o
[ 37%] Linking CXX executable ../../../bin/runQCProducerDevice
[ 37%] Built target runQCProducerDevice
[ 37%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/DeviceSpecHelpers.cxx.o
[ 37%] Linking CXX shared library ../../lib/libaliceHLTwrapper.so
[ 37%] Built target aliceHLTwrapper
[ 37%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/DDSConfigHelpers.cxx.o
[ 38%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/FairOptionsRetriever.cxx.o
[ 38%] Linking CXX executable ../../../bin/runQCMergerDevice
[ 38%] Built target runQCMergerDevice
Scanning dependencies of target runQCViewerDevice
[ 38%] Building CXX object Utilities/QC/QCViewer/CMakeFiles/runQCViewerDevice.dir/src/runViewerDevice.cxx.o
[ 38%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/GraphvizHelpers.cxx.o
[ 38%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/InputRecord.cxx.o
Scanning dependencies of target runQCMetricsExtractor
[ 38%] Building CXX object Utilities/QC/QCMetricsExtractor/CMakeFiles/runQCMetricsExtractor.dir/src/runMetricsExtractor.cxx.o
Scanning dependencies of target DataFlow
[ 38%] Building CXX object Utilities/DataFlow/CMakeFiles/DataFlow.dir/src/FakeTimeframeBuilder.cxx.o
[ 38%] Linking CXX executable ../../../bin/runQCViewerDevice
[ 38%] Built target runQCViewerDevice
[ 38%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/LocalRootFileService.cxx.o
[ 38%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/LogParsingHelpers.cxx.o
[ 38%] Linking CXX executable ../../../bin/runQCMetricsExtractor
[ 38%] Building CXX object Utilities/DataFlow/CMakeFiles/DataFlow.dir/src/FakeTimeframeGeneratorDevice.cxx.o
[ 38%] Built target runQCMetricsExtractor
Scanning dependencies of target Publishers
[ 38%] Building CXX object Utilities/Publishers/CMakeFiles/Publishers.dir/src/DataPublisherDevice.cxx.o
[ 38%] Building CXX object Utilities/DataFlow/CMakeFiles/DataFlow.dir/src/HeartbeatSampler.cxx.o
[ 38%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/PaletteHelpers.cxx.o
In file included from /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DAQ/0/Framework/Core/src/PaletteHelpers.cxx:10:0:
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DAQ/0/Framework/Core/include/Framework/PaletteHelpers.h:13:28: fatal error: DebugGUI/imgui.h: No such file or directory
 #include "DebugGUI/imgui.h"
                            ^
compilation terminated.
make[2]: *** [Framework/Core/CMakeFiles/Framework.dir/src/PaletteHelpers.cxx.o] Error 1
make[1]: *** [Framework/Core/CMakeFiles/Framework.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 38%] Building CXX object Utilities/DataFlow/CMakeFiles/DataFlow.dir/src/SubframeBuilderDevice.cxx.o
[ 38%] Building CXX object Utilities/DataFlow/CMakeFiles/DataFlow.dir/src/TimeframeParser.cxx.o
[ 39%] Building CXX object Utilities/DataFlow/CMakeFiles/DataFlow.dir/src/TimeframeReaderDevice.cxx.o
[ 39%] Linking CXX shared library ../../lib/libPublishers.so
[ 39%] Built target Publishers
[ 39%] Building CXX object Utilities/DataFlow/CMakeFiles/DataFlow.dir/src/TimeframeValidatorDevice.cxx.o
[ 39%] Building CXX object Utilities/DataFlow/CMakeFiles/DataFlow.dir/src/TimeframeWriterDevice.cxx.o
[ 39%] Building CXX object Utilities/DataFlow/CMakeFiles/DataFlow.dir/src/EPNReceiverDevice.cxx.o
[ 39%] Building CXX object Utilities/DataFlow/CMakeFiles/DataFlow.dir/src/FLPSenderDevice.cxx.o
[ 39%] Linking CXX shared library ../../lib/libDataFlow.so
[ 39%] Built target DataFlow
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/742/0e2adac8e237c59b58c9f59802920d1c8ca1eae0/fullLog.txt).

alibuild commented:
Error while checking build/O2/o2 for 0e2adac8e237c59b58c9f59802920d1c8ca1eae0:
```
sw/BUILD/O2-latest/log
[ 26%] Building CXX object Utilities/aliceHLTwrapper/CMakeFiles/aliceHLTwrapper.dir/src/WrapperDevice.cxx.o
[ 26%] Building CXX object Utilities/aliceHLTwrapper/CMakeFiles/aliceHLTwrapper.dir/src/Component.cxx.o
[ 26%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/DataAllocator.cxx.o
[ 27%] Linking CXX executable ../../bin/epnReceiver
[ 27%] Built target epnReceiver
[ 27%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/DataProcessingDevice.cxx.o
[ 27%] Building CXX object Utilities/aliceHLTwrapper/CMakeFiles/aliceHLTwrapper.dir/src/MessageFormat.cxx.o
[ 27%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/DataProcessingHeader.cxx.o
Scanning dependencies of target runO2MessageMonitor
[ 27%] Building CXX object Utilities/O2MessageMonitor/CMakeFiles/runO2MessageMonitor.dir/src/runO2MessageMonitor.cxx.o
[ 28%] Building CXX object Utilities/aliceHLTwrapper/CMakeFiles/aliceHLTwrapper.dir/src/EventSampler.cxx.o
[ 28%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/DataProcessor.cxx.o
[ 28%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/DataRelayer.cxx.o
[ 29%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/DataSourceDevice.cxx.o
[ 29%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/DeviceMetricsInfo.cxx.o
[ 29%] Linking CXX executable ../../bin/runO2MessageMonitor
[ 29%] Built target runO2MessageMonitor
[ 29%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/DeviceSpec.cxx.o
[ 29%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/DeviceSpecHelpers.cxx.o
[ 29%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/DDSConfigHelpers.cxx.o
[ 29%] Linking CXX shared library ../../lib/libaliceHLTwrapper.so
[ 29%] Built target aliceHLTwrapper
[ 29%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/FairOptionsRetriever.cxx.o
[ 30%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/GraphvizHelpers.cxx.o
Scanning dependencies of target runQCProducerDevice
[ 30%] Building CXX object Utilities/QC/QCProducer/CMakeFiles/runQCProducerDevice.dir/src/runProducer.cxx.o
[ 30%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/InputRecord.cxx.o
[ 30%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/LocalRootFileService.cxx.o
[ 30%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/LogParsingHelpers.cxx.o
[ 30%] Building CXX object Framework/Core/CMakeFiles/Framework.dir/src/PaletteHelpers.cxx.o
In file included from /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/Framework/Core/src/PaletteHelpers.cxx:10:0:
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/Framework/Core/include/Framework/PaletteHelpers.h:13:28: fatal error: DebugGUI/imgui.h: No such file or directory
 #include "DebugGUI/imgui.h"
                            ^
compilation terminated.
make[2]: *** [Framework/Core/CMakeFiles/Framework.dir/src/PaletteHelpers.cxx.o] Error 1
make[2]: *** Waiting for unfinished jobs....
Scanning dependencies of target runQCMergerDevice
[ 30%] Building CXX object Utilities/QC/QCMerger/CMakeFiles/runQCMergerDevice.dir/src/runMerger.cxx.o
Scanning dependencies of target runQCViewerDevice
[ 30%] Building CXX object Utilities/QC/QCViewer/CMakeFiles/runQCViewerDevice.dir/src/runViewerDevice.cxx.o
make[1]: *** [Framework/Core/CMakeFiles/Framework.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 30%] Linking CXX executable ../../../bin/runQCProducerDevice
[ 30%] Built target runQCProducerDevice
[ 30%] Linking CXX executable ../../../bin/runQCViewerDevice
[ 30%] Built target runQCViewerDevice
[ 30%] Linking CXX executable ../../../bin/runQCMergerDevice
[ 30%] Built target runQCMergerDevice
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/742/0e2adac8e237c59b58c9f59802920d1c8ca1eae0/fullLog.txt).

mkrzewic commented on Framework/Core/src/runDataProcessing.cxx:
```diff
@@ -311,20 +325,32 @@ int createPipes(int maxFd, int *pipes) {
+    if (exitCode == 0
+        && info.maxLogLevel >= LogParsingHelpers::LogLevel::Error){
+      exitCode = 1;
+    }
```
actually, why use return codes for error conditions? would the use of exceptions not be clearer? This I mean in general, i have seen quite some code in O2 already that uses retCodes and/or assert() - should we not try to push it in a consistent direction?

ktf replied:
This is not an error condition. It's the calculated value which the driver will use to call `exit()`. So I would say it does not apply here. 

That said, I would personally avoid at all costs exceptions for anything but things which requires exiting the computation immediately and discarding the data.

alibuild commented:
Error while checking build/o2checkcode/o2 for 668e7a76b0166607a51b4f511f7e8a7d422e0b4e:
```
sw/BUILD/O2-latest/log
      Start 35: test_Framework_test_DeviceSpec
      Start 38: test_Framework_test_InputRecord
      Start 55: test_aliceHLTwrapper_testMessageFormat
45/65 Test #54: test_Algorithm_pageparser ..........................   Passed    0.66 sec
46/65 Test #35: test_Framework_test_DeviceSpec .....................   Passed    0.52 sec
47/65 Test #55: test_aliceHLTwrapper_testMessageFormat .............   Passed    0.53 sec
      Start 33: test_Framework_test_DataRefUtils
      Start 43: test_Framework_test_SingleDataSource
      Start 58: test_DataFlow_test_SubframeUtils01
48/65 Test #38: test_Framework_test_InputRecord ....................   Passed    0.54 sec
      Start  6: test_Headers_testTimeStamp
49/65 Test #58: test_DataFlow_test_SubframeUtils01 .................   Passed    0.52 sec
50/65 Test #43: test_Framework_test_SingleDataSource ...............   Passed    0.52 sec
      Start 31: test_Framework_test_BoostOptionsRetriever
      Start 50: test_Algorithm_o2formatparser
51/65 Test #33: test_Framework_test_DataRefUtils ...................   Passed    0.62 sec
      Start 65: test_Steer_testInteractionSampler
52/65 Test  #6: test_Headers_testTimeStamp .........................   Passed    0.54 sec
53/65 Test #31: test_Framework_test_BoostOptionsRetriever ..........   Passed    0.51 sec
54/65 Test #50: test_Algorithm_o2formatparser ......................   Passed    0.51 sec
      Start 63: test_DataCompression_test_HuffmanCodec
      Start  7: test_Headers_test_HeartbeatFrame
      Start 19: test_TPCBase_testTPCParameters
55/65 Test #65: test_Steer_testInteractionSampler ..................   Passed    0.72 sec
56/65 Test #63: test_DataCompression_test_HuffmanCodec .............   Passed    0.51 sec
57/65 Test  #7: test_Headers_test_HeartbeatFrame ...................   Passed    0.50 sec
      Start 52: test_Algorithm_parser
      Start  5: test_Headers_testDataHeader
      Start  8: test_Headers_test_RAWDataHeader
58/65 Test #19: test_TPCBase_testTPCParameters .....................   Passed    0.61 sec
      Start 64: test_DataCompression_test_DataDeflater
59/65 Test #64: test_DataCompression_test_DataDeflater .............   Passed    0.10 sec
60/65 Test #52: test_Algorithm_parser ..............................   Passed    0.48 sec
61/65 Test  #5: test_Headers_testDataHeader ........................   Passed    0.49 sec
      Start 29: test_ExampleModule1_testExampleModule1
      Start 60: test_DataCompression_test_dc_primitives
      Start 40: test_Framework_test_LogParsingHelpers
62/65 Test  #8: test_Headers_test_RAWDataHeader ....................   Passed    0.50 sec
63/65 Test #29: test_ExampleModule1_testExampleModule1 .............   Passed    0.11 sec
64/65 Test #60: test_DataCompression_test_dc_primitives ............   Passed    0.11 sec
65/65 Test #40: test_Framework_test_LogParsingHelpers ..............   Passed    0.49 sec

97% tests passed, 2 tests failed out of 65

Total Test time (real) =  16.40 sec

The following tests FAILED:
	 39 - test_Framework_test_ParallelProducer (Failed)
	 49 - test_Framework_test_WorkflowOptions (Failed)
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/742/668e7a76b0166607a51b4f511f7e8a7d422e0b4e/fullLog.txt).

alibuild commented:
Error while checking build/O2/o2 for 668e7a76b0166607a51b4f511f7e8a7d422e0b4e:
```
sw/BUILD/O2-latest/log
118/139 Test  #40: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/Detectors/gconfig/DecayConfig.C ..................................   Passed    0.42 sec
        Start 136: test_DataCompression_test_HuffmanCodec
        Start  78: test_Headers_test_HeartbeatFrame
119/139 Test #105: test_Framework_test_DataRelayer ....................................................................................   Passed    0.83 sec
        Start 130: test_DataFlow_test_TimeframeParser
120/139 Test  #73: test_MathUtils_testCartesian3D .....................................................................................   Passed    0.54 sec
        Start  42: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/Detectors/gconfig/UserDecay.C
121/139 Test #136: test_DataCompression_test_HuffmanCodec .............................................................................   Passed    0.62 sec
122/139 Test  #78: test_Headers_test_HeartbeatFrame ...................................................................................   Passed    0.71 sec
        Start  77: test_Headers_testTimeStamp
        Start  41: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/Detectors/gconfig/SetCuts.C
123/139 Test #130: test_DataFlow_test_TimeframeParser .................................................................................   Passed    0.72 sec
        Start  50: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/macro/SetIncludePath.C
124/139 Test  #42: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/Detectors/gconfig/UserDecay.C ....................................   Passed    0.51 sec
125/139 Test  #41: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/Detectors/gconfig/SetCuts.C ......................................   Passed    0.39 sec
        Start  10: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/Detectors/ITSMFT/ITS/macros/test/rootlogon.C
        Start  45: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/Detectors/gconfig/g3libs.C
126/139 Test  #77: test_Headers_testTimeStamp .........................................................................................   Passed    0.80 sec
127/139 Test  #50: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/macro/SetIncludePath.C ...........................................   Passed    0.49 sec
        Start  13: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/Detectors/ITSMFT/MFT/macros/SetIncludePath.C
        Start  28: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/Detectors/TPC/reconstruction/macro/addInclude.C
128/139 Test  #10: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/Detectors/ITSMFT/ITS/macros/test/rootlogon.C .....................   Passed    0.41 sec
129/139 Test  #28: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/Detectors/TPC/reconstruction/macro/addInclude.C ..................   Passed    0.25 sec
        Start  14: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/Detectors/ITSMFT/MFT/macros/load_all_libs.C
        Start  15: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/Detectors/ITSMFT/MFT/macros/rootlogon.C
130/139 Test  #45: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/Detectors/gconfig/g3libs.C .......................................   Passed    0.46 sec
131/139 Test  #13: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/Detectors/ITSMFT/MFT/macros/SetIncludePath.C .....................   Passed    0.33 sec
132/139 Test  #15: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/Detectors/ITSMFT/MFT/macros/rootlogon.C ..........................   Passed    0.26 sec
        Start  56: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/macro/rootlogon.C
        Start  53: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/macro/load_all_libs.C
        Start 102: test_ExampleModule1_testExampleModule1
133/139 Test #102: test_ExampleModule1_testExampleModule1 .............................................................................   Passed    0.10 sec
134/139 Test  #14: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/Detectors/ITSMFT/MFT/macros/load_all_libs.C ......................   Passed    0.47 sec
        Start 133: test_DataCompression_test_dc_primitives
        Start 137: test_DataCompression_test_DataDeflater
135/139 Test #137: test_DataCompression_test_DataDeflater .............................................................................   Passed    0.01 sec
136/139 Test  #56: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/macro/rootlogon.C ................................................   Passed    0.27 sec
137/139 Test  #53: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/master/0/macro/load_all_libs.C ............................................   Passed    0.28 sec
138/139 Test #133: test_DataCompression_test_dc_primitives ............................................................................   Passed    0.07 sec
        Start 113: test_Framework_test_LogParsingHelpers
139/139 Test #113: test_Framework_test_LogParsingHelpers ..............................................................................   Passed    0.55 sec

99% tests passed, 2 tests failed out of 139

Total Test time (real) =  82.15 sec

The following tests FAILED:
	112 - test_Framework_test_ParallelProducer (Failed)
	122 - test_Framework_test_WorkflowOptions (Failed)
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/742/668e7a76b0166607a51b4f511f7e8a7d422e0b4e/fullLog.txt).

alibuild commented:
Error while checking build/O2/o2-dev-fairroot for 668e7a76b0166607a51b4f511f7e8a7d422e0b4e:
```
sw/BUILD/O2-latest/log
        Start  87: test_TPCBase_testTPCBase
118/139 Test #123: test_Algorithm_o2formatparser .................................................................................................   Passed    0.60 sec
        Start 139: test_Steer_testInteractionSampler
119/139 Test #121: test_Framework_test_WorkflowHelpers ...........................................................................................   Passed    1.00 sec
120/139 Test #117: test_Framework_test_SuppressionGenerator ......................................................................................   Passed    0.80 sec
121/139 Test  #87: test_TPCBase_testTPCBase ......................................................................................................   Passed    0.53 sec
        Start  77: test_Headers_testTimeStamp
        Start 130: test_DataFlow_test_TimeframeParser
        Start  50: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/macro/SetIncludePath.C
122/139 Test #139: test_Steer_testInteractionSampler .............................................................................................   Passed    0.64 sec
123/139 Test  #50: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/macro/SetIncludePath.C ...........................................   Passed    0.50 sec
124/139 Test  #77: test_Headers_testTimeStamp ....................................................................................................   Passed    0.64 sec
        Start  78: test_Headers_test_HeartbeatFrame
        Start  13: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/Detectors/ITSMFT/MFT/macros/SetIncludePath.C
        Start  41: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/Detectors/gconfig/SetCuts.C
125/139 Test #130: test_DataFlow_test_TimeframeParser ............................................................................................   Passed    0.84 sec
        Start  56: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/macro/rootlogon.C
126/139 Test  #13: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/Detectors/ITSMFT/MFT/macros/SetIncludePath.C .....................   Passed    0.61 sec
        Start  45: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/Detectors/gconfig/g3libs.C
127/139 Test  #78: test_Headers_test_HeartbeatFrame ..............................................................................................   Passed    0.69 sec
128/139 Test  #41: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/Detectors/gconfig/SetCuts.C ......................................   Passed    0.69 sec
129/139 Test  #56: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/macro/rootlogon.C ................................................   Passed    0.48 sec
        Start  10: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/Detectors/ITSMFT/ITS/macros/test/rootlogon.C
        Start  15: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/Detectors/ITSMFT/MFT/macros/rootlogon.C
        Start  14: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/Detectors/ITSMFT/MFT/macros/load_all_libs.C
130/139 Test  #10: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/Detectors/ITSMFT/ITS/macros/test/rootlogon.C .....................   Passed    0.48 sec
131/139 Test  #45: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/Detectors/gconfig/g3libs.C .......................................   Passed    0.66 sec
132/139 Test  #15: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/Detectors/ITSMFT/MFT/macros/rootlogon.C ..........................   Passed    0.47 sec
133/139 Test  #14: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/Detectors/ITSMFT/MFT/macros/load_all_libs.C ......................   Passed    0.47 sec
        Start  28: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/Detectors/TPC/reconstruction/macro/addInclude.C
        Start  53: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/macro/load_all_libs.C
        Start 102: test_ExampleModule1_testExampleModule1
        Start 137: test_DataCompression_test_DataDeflater
134/139 Test #102: test_ExampleModule1_testExampleModule1 ........................................................................................   Passed    0.10 sec
135/139 Test #137: test_DataCompression_test_DataDeflater ........................................................................................   Passed    0.20 sec
        Start 133: test_DataCompression_test_dc_primitives
        Start 113: test_Framework_test_LogParsingHelpers
136/139 Test #133: test_DataCompression_test_dc_primitives .......................................................................................   Passed    0.10 sec
137/139 Test  #28: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/Detectors/TPC/reconstruction/macro/addInclude.C ..................   Passed    0.51 sec
138/139 Test  #53: /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DEV_FAIRROOT/0/macro/load_all_libs.C ............................................   Passed    0.51 sec
139/139 Test #113: test_Framework_test_LogParsingHelpers .........................................................................................   Passed    0.55 sec

99% tests passed, 2 tests failed out of 139

Total Test time (real) =  77.91 sec

The following tests FAILED:
	112 - test_Framework_test_ParallelProducer (Failed)
	122 - test_Framework_test_WorkflowOptions (Failed)
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/742/668e7a76b0166607a51b4f511f7e8a7d422e0b4e/fullLog.txt).

alibuild commented:
Error while checking build/o2/macos for 668e7a76b0166607a51b4f511f7e8a7d422e0b4e:
```
sw/BUILD/O2-latest/log
[[01;36m23:29:58[0m][[01;32mINFO[0m] Found control command QUIT valid for ALL
[[01;36m23:29:58[0m][[01;32mINFO[0m] Found control command QUIT valid for ALL
[[01;36m23:29:58[0m][[01;32mINFO[0m] Found control command QUIT valid for ALL
[[01;36m23:29:58[0m][[01;32mINFO[0m] Found control command QUIT valid for ALL
[[01;36m23:29:58[0m][[01;32mINFO[0m] Found control command QUIT valid for ALL
[[01;36m23:29:58[0m][[01;32mINFO[0m] Found control command QUIT valid for ALL
[[01;36m23:29:58[0m][[01;32mINFO[0m] Found control command QUIT valid for ALL
[[01;36m23:29:58[0m][[01;32mINFO[0m] Found control command QUIT valid for ALL
[[01;36m23:29:58[0m][[01;32mINFO[0m] Found control command QUIT valid for ALL
[[01;36m23:29:58[0m][[01;32mINFO[0m] Found control command QUIT valid for ALL
[[01;36m23:29:58[0m][[01;32mINFO[0m] Found control command QUIT valid for ALL
[54615]: CONT    = sub                                                                                                 <string>         [provided]
Child exited: 54615
Child exited: 54616
Process 54609 is exiting.

        Start 130: test_DataFlow_test_TimeframeParser
        Start 131: test_DataFlow_test_SubframeUtils01
        Start 132: test_DataFlow_test_PayloadMerger01
125/139 Test #127: test_Algorithm_pageparser ......................................................................................................   Passed    1.38 sec
        Start 133: test_DataCompression_test_dc_primitives
126/139 Test #128: test_aliceHLTwrapper_testMessageFormat .........................................................................................   Passed    2.02 sec
127/139 Test #129: test_O2MessageMonitor_O2MessageMonitorTest .....................................................................................   Passed    2.01 sec
        Start 134: test_DataCompression_test_Fifo
        Start 135: test_DataCompression_test_DataGenerator
128/139 Test #131: test_DataFlow_test_SubframeUtils01 .............................................................................................   Passed    1.78 sec
129/139 Test #133: test_DataCompression_test_dc_primitives ........................................................................................   Passed    0.94 sec
130/139 Test #130: test_DataFlow_test_TimeframeParser .............................................................................................   Passed    1.79 sec
131/139 Test #132: test_DataFlow_test_PayloadMerger01 .............................................................................................   Passed    1.78 sec
        Start 136: test_DataCompression_test_HuffmanCodec
        Start 137: test_DataCompression_test_DataDeflater
        Start 138: mcloggertest
        Start 139: test_Steer_testInteractionSampler
132/139 Test #137: test_DataCompression_test_DataDeflater .........................................................................................   Passed    0.38 sec
133/139 Test #139: test_Steer_testInteractionSampler ..............................................................................................   Passed    1.60 sec
134/139 Test #136: test_DataCompression_test_HuffmanCodec .........................................................................................   Passed    1.82 sec
135/139 Test #135: test_DataCompression_test_DataGenerator ........................................................................................   Passed    3.21 sec
136/139 Test  #96: test_TPCSimulation_testTPCGEMAmplification .....................................................................................   Passed   21.42 sec
137/139 Test #101: test_TOFBase_testTOFIndex ......................................................................................................   Passed   19.06 sec
138/139 Test #138: mcloggertest ...................................................................................................................   Passed    5.40 sec
139/139 Test #134: test_DataCompression_test_Fifo .................................................................................................   Passed    7.29 sec

99% tests passed, 2 tests failed out of 139

Total Test time (real) = 125.96 sec

The following tests FAILED:
	112 - test_Framework_test_ParallelProducer (Failed)
	122 - test_Framework_test_WorkflowOptions (Failed)
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/742/668e7a76b0166607a51b4f511f7e8a7d422e0b4e/fullLog.txt).

matthiasrichter commented on Framework/Core/CMakeLists.txt:
```diff
@@ -15,7 +15,8 @@ set(MODULE_BUCKET_NAME O2FrameworkCore_bucket)
   set(GUI_SOURCES src/FrameworkGUIDebugger.cxx
-                  src/FrameworkGUIDevicesGraph.cxx)
+                  src/FrameworkGUIDevicesGraph.cxx
+                  src/PaletteHelpers.cxx)
```
Since there is a change here which draws my attention: at some point we should move the debug GUI to a different folder than `Core`

ktf replied:
Yes, agreed, but I'd rather do it after I see how to do it with @dennisklein changes to the CMake stuff.

matthiasrichter commented on Framework/Core/include/Framework/DeviceControl.h:
```diff
@@ -12,21 +12,30 @@
+  /// Minimum log level for messages to appear 
+  LogParsingHelpers::LogLevel logLevel = LogParsingHelpers::LogLevel::Info;
+  /// Lines in the log should match this to be displayed  
+  char logFilter[256] = {0};
```
not really touched by this PR, but lets define a constant holding the length. I found a couple of `asserts` checking the size of the `log*` members to be 256 in `FrameworkGUIDebugger.cxx`.

Also I noticed, in `FrameworkGUIDebugger.cxx` line 367
```
ImGui::InputText("Log filter", control.logFilter, sizeof(control.logFilter));
```
there is no check that the array is 0 terminated.

ktf replied:
I think it does not need to be 0 terminated. In the example the author of ImGUI however uses `IM_ARRAYSIZE`, so I will do the same.

matthiasrichter commented on Framework/Core/test/test_LogParsingHelpers.cxx:
```diff
@@ -0,0 +1,29 @@
+
+#include "Framework/LogParsingHelpers.h"
+#include <boost/test/unit_test.hpp>
+#include <regex>
```
remnant from testing?

ktf replied:
No, not really, I do want to make sure I have a test which fails if the log format changes and I adapt `parseTokenLevel`.

matthiasrichter commented on Framework/Core/src/FrameworkGUIDebugger.cxx:
```diff
@@ -320,6 +348,45 @@ struct ChannelsTableHelper {
+  if (info.active == false) {
+    ImGui::PushStyleColor(ImGuiCol_TitleBg, ImVec4(0.9,0,0,1));
+    ImGui::PushStyleColor(ImGuiCol_TitleBgActive, ImVec4(1,0,0,1));
+    ImGui::PushStyleColor(ImGuiCol_TitleBgCollapsed, ImVec4(1,0,0,1));
```
define color aliases for these as well?

ktf replied:
Agreed. Will change this.

matthiasrichter commented on Framework/Core/src/FrameworkGUIDebugger.cxx:
```diff
@@ -74,6 +76,24 @@ optionsTable(const DeviceSpec &spec, const DeviceControl &control) {
+    case LogParsingHelpers::LogLevel::Warning:
+      return PaletteHelpers::DARK_YELLOW;
+    case LogParsingHelpers::LogLevel::Error:
+      return PaletteHelpers::DARK_RED;
```
just noticed that this is difficult to read on the black background, would change to `RED`

ktf replied:
I will change this.

ktf commented:
Summarizing what I need to do:

- [ ] Change to RED the colour for the errors.
- [ ] Change to IM_ARRAYSIZE for ImGui::InputText
- [ ] Make 256 a constant
- [ ] Add a few more colours for the palette and use them rather than hardcoded values.

alibuild commented:
Error while checking build/o2/macos for 53963892866d7bfcc263fa35784af3dc1bc0f6e2:
```
sw/BUILD/O2-latest/log
        Start 136: test_DataFlow_test_SubframeUtils01
129/144 Test #127: test_Framework_test_WorkflowOptions ..................................................................................................***Failed    3.48 sec
[[01;36m14:59:02[0m][[01;34mDEBUG[0m] The following options are being forwarded to simple: /build/ci_checks/o2_checker_macos_o2_ci_0/sw/BUILD/e6a343d86418227bb87a1d3648408aa82e4fe49d/O2/bin/test_Framework_test_WorkflowOptions --id simple --control static --log-color 0 --channel-config name=foo,type=sub,method=connect,address=tcp://localhost:5450,rateLogging=1 --channel-config name=from_simple_to_simple2,type=pub,method=bind,address=tcp://*:22000
[[01;36m14:59:02[0m][[01;34mDEBUG[0m] The following options are being forwarded to simple2: /build/ci_checks/o2_checker_macos_o2_ci_0/sw/BUILD/e6a343d86418227bb87a1d3648408aa82e4fe49d/O2/bin/test_Framework_test_WorkflowOptions --id simple2 --control static --log-color 0 --channel-config name=from_simple_to_simple2,type=sub,method=connect,address=tcp://127.0.0.1:22000
[[01;36m14:59:02[0m][[01;33mWARN[0m] Could not create GUI. Switching to batch mode. Do you have GLFW on your system?
Starting simple on pid 90409
Starting simple2 on pid 90410
[90409]: [14:59:04][INFO] Spawing new device simple in process with pid 90409
[90409]: 
[90409]: [14:59:04][STATE] Starting FairMQ state machine
[[01;36m14:59:04[0m][[01;32mINFO[0m] Found control command QUIT valid for ALL
[[01;36m14:59:04[0m][[01;32mINFO[0m] Found control command QUIT valid for ALL
[[01;36m14:59:04[0m][[01;32mINFO[0m] Found control command QUIT valid for ALL
[90409]: le2.0.method        = bind                                                                                                <string>         [provided]
Child exited: 90409
Child exited: 90410
Process 90406 is exiting.

130/144 Test #133: test_aliceHLTwrapper_testMessageFormat ...............................................................................................   Passed    1.69 sec
131/144 Test #132: test_Algorithm_pageparser ............................................................................................................   Passed    1.69 sec
        Start 137: test_DataFlow_test_PayloadMerger01
        Start 138: test_DataCompression_test_dc_primitives
        Start 139: test_DataCompression_test_Fifo
132/144 Test #134: test_O2MessageMonitor_O2MessageMonitorTest ...........................................................................................   Passed    1.38 sec
133/144 Test #138: test_DataCompression_test_dc_primitives ..............................................................................................   Passed    0.42 sec
        Start 140: test_DataCompression_test_DataGenerator
        Start 141: test_DataCompression_test_HuffmanCodec
134/144 Test #135: test_DataFlow_test_TimeframeParser ...................................................................................................   Passed    1.28 sec
135/144 Test #136: test_DataFlow_test_SubframeUtils01 ...................................................................................................   Passed    1.88 sec
        Start 142: test_DataCompression_test_DataDeflater
        Start 143: mcloggertest
136/144 Test #137: test_DataFlow_test_PayloadMerger01 ...................................................................................................   Passed    1.77 sec
        Start 144: test_Steer_testInteractionSampler
137/144 Test #142: test_DataCompression_test_DataDeflater ...............................................................................................   Passed    1.05 sec
138/144 Test #101: test_TPCSimulation_testTPCGEMAmplification ...........................................................................................   Passed   19.91 sec
139/144 Test #144: test_Steer_testInteractionSampler ....................................................................................................   Passed    1.44 sec
140/144 Test #140: test_DataCompression_test_DataGenerator ..............................................................................................   Passed    3.10 sec
141/144 Test #106: test_TOFBase_testTOFIndex ............................................................................................................   Passed   20.42 sec
142/144 Test #143: mcloggertest .........................................................................................................................   Passed    6.11 sec
143/144 Test #139: test_DataCompression_test_Fifo .......................................................................................................   Passed    7.46 sec
144/144 Test #141: test_DataCompression_test_HuffmanCodec ...............................................................................................   Passed   15.13 sec

99% tests passed, 2 tests failed out of 144

Total Test time (real) = 166.21 sec

The following tests FAILED:
	117 - test_Framework_test_ParallelProducer (Failed)
	127 - test_Framework_test_WorkflowOptions (Failed)
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/742/53963892866d7bfcc263fa35784af3dc1bc0f6e2/fullLog.txt).

ktf commented:
@matthiasrichter @sawenzel Tests will disprove me, but I think I addressed all the comments made and fixed the issue with the two tests failing. Unless there is some strong objection to the content of this PR, could you please approve it?

alibuild commented:
Error while checking build/o2/macos for b1db90089eb883a9b5aacae1398ed07846734a3f:
```
sw/BUILD/O2-latest/log
121/144 Test #110: test_Framework_test_DataRelayer ......................................................................................................   Passed    2.72 sec
        Start 132: test_Algorithm_pageparser
        Start 126: test_Framework_test_WorkflowHelpers
122/144 Test  #81: test_Headers_testDataHeader ..........................................................................................................   Passed    2.27 sec
123/144 Test #137: test_DataFlow_test_PayloadMerger01 ...................................................................................................   Passed    2.25 sec
124/144 Test #136: test_DataFlow_test_SubframeUtils01 ...................................................................................................   Passed    2.26 sec
125/144 Test  #97: test_TPCReconstruction_testTPCAdcClockMonitor ........................................................................................   Passed    2.33 sec
        Start 133: test_aliceHLTwrapper_testMessageFormat
        Start  96: test_TPCReconstruction_testTPCSyncPatternMonitor
        Start  92: test_TPCBase_testTPCBase
        Start 128: test_Algorithm_o2formatparser
126/144 Test  #84: test_Headers_test_RAWDataHeader ......................................................................................................   Passed    2.22 sec
127/144 Test #118: test_Framework_test_LogParsingHelpers ................................................................................................   Passed    2.53 sec
        Start 112: test_Framework_test_DeviceMetricsInfo
        Start  90: test_DetectorsBase_testDetID
128/144 Test #132: test_Algorithm_pageparser ............................................................................................................   Passed    2.20 sec
129/144 Test #126: test_Framework_test_WorkflowHelpers ..................................................................................................   Passed    2.30 sec
        Start 144: test_Steer_testInteractionSampler
        Start 129: test_Algorithm_headerstack
130/144 Test #133: test_aliceHLTwrapper_testMessageFormat ...............................................................................................   Passed    2.22 sec
131/144 Test  #92: test_TPCBase_testTPCBase .............................................................................................................   Passed    2.24 sec
132/144 Test  #96: test_TPCReconstruction_testTPCSyncPatternMonitor .....................................................................................   Passed    2.45 sec
133/144 Test #128: test_Algorithm_o2formatparser ........................................................................................................   Passed    2.55 sec
        Start 134: test_O2MessageMonitor_O2MessageMonitorTest
        Start 124: test_Framework_test_TypeTraits
        Start 135: test_DataFlow_test_TimeframeParser
        Start 122: test_Framework_test_SuppressionGenerator
134/144 Test #112: test_Framework_test_DeviceMetricsInfo ................................................................................................   Passed    2.26 sec
135/144 Test  #90: test_DetectorsBase_testDetID .........................................................................................................   Passed    2.46 sec
        Start 142: test_DataCompression_test_DataDeflater
        Start 107: test_ExampleModule1_testExampleModule1
136/144 Test #142: test_DataCompression_test_DataDeflater ...............................................................................................   Passed    0.65 sec
        Start 138: test_DataCompression_test_dc_primitives
137/144 Test #144: test_Steer_testInteractionSampler ....................................................................................................   Passed    2.09 sec
138/144 Test #107: test_ExampleModule1_testExampleModule1 ...............................................................................................   Passed    0.88 sec
139/144 Test #129: test_Algorithm_headerstack ...........................................................................................................   Passed    2.51 sec
140/144 Test #122: test_Framework_test_SuppressionGenerator .............................................................................................   Passed    2.11 sec
141/144 Test #134: test_O2MessageMonitor_O2MessageMonitorTest ...........................................................................................   Passed    2.18 sec
142/144 Test #138: test_DataCompression_test_dc_primitives ..............................................................................................   Passed    0.77 sec
143/144 Test #135: test_DataFlow_test_TimeframeParser ...................................................................................................   Passed    2.17 sec
144/144 Test #124: test_Framework_test_TypeTraits .......................................................................................................   Passed    2.17 sec

99% tests passed, 2 tests failed out of 144

Total Test time (real) = 207.35 sec

The following tests FAILED:
	117 - test_Framework_test_ParallelProducer (Failed)
	127 - test_Framework_test_WorkflowOptions (Failed)
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/742/b1db90089eb883a9b5aacae1398ed07846734a3f/fullLog.txt).

ktf commented:
@brinick @dberzano do we have issues with the O2 checkers on linux?

ktf commented:
All tests pass and the feature is rather unambiguous. I will merge this so that I can start integrating my next set of (unrelated) changes. Of course we can always revisit the behavior and implementation if needed / requested.
